zipl/man/zipl-editenv.8.in
.\" Copyright 2020 IBM Corp.
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\"
.TH ZIPL-EDITENV 8 "Feb 2020" "s390-tools"
.SH NAME
zipl-editenv \- zIPL environment editor.

.SH SYNOPSIS
.B zipl-editenv
[OPTION] [ARGUMENT]

.SH DESCRIPTION
.B zipl-editenv
is a tool to manipulate with zIPL environment of partitions, prepared for
initial program load (IPL) by
.B zipl(8)
utility. This tool performs atomic updates of boot data without modifying the
boot record referencing that data. The following operations are supported:
.IP "     -"
Print environment
.IP "     -"
Assign a specified value to any environment variable
.IP "     -"
Remove a specified variable from the environment
.IP "     -"
Remove all variables from the environment
.PP

.B zIPL environment
(don't confuse with shell environment!) is a method of evaluation of zIPL
environment variables, which is used at boot time. Names of zIPL environment
variables have to satisfy POSIX requirements for shell environment variables
(IEEE Std 1003.1-2001). That is, such names should consist solely of uppercase
letters, digits, and underscores ('_') from the characters defined in Portable
Character Set and should not begin with a digit.

zIPL environment is defined per boot partition by a special boot component
called
.B zIPL environment block.
This component is used for evaluation of zIPL environment variables present
in another boot component, kernel command line, where the mentioned variables
have to be specified by their names as ${NAME}. During boot, such variables
are replaced with values, as defined by the active namespace in the installed
environment block.

zIPL environment is installed every time when preparing a device for initial
program load (IPL) by
.B zipl(8)
utility. At installation time the environment by default is imported from
user-editable file
.B '/etc/ziplenv'.
That default location can be overridden by respective
.B zipl(8)
option.

Installed zIPL environment (AKA zIPL environment block) can be accessed
(modified, printed) by this
.B zipl-editenv(8)
tool. All modifications made with this tool don't affect the file
.B '/etc/ziplenv'.
Hence, those modifications will be lost after the next boot record
re-installation, unless
.B '/etc/ziplenv'
was updated respectively.

At boot time any zIPL environment variable can be evaluated by different
preinstalled ways. An individual method of evaluation of all environment
variables is called an
.B environment namespace.
zIPL supports 10 identified namespaces called
.B sites
and one so-called
.B common
(or default) namespace per boot partition.
Maximum number of values defined by all installed namespaces is 512.
Sites are identified by positive integer numbers from 0 to 9.
In contrast with sites, common namespace doesn't have any identifier.
User can set anyone of the installed site namespaces to be used for
evaluation of zIPL environment variables (in the kernel command line)
at boot time. This operation is also called
.B namespace activation.
Namespace can be activated either "in advance" by
.B chreipl(8)
utility for the next boot session, or directly at boot time by the boot
command. In both ways ID of the active site should be properly encoded
in the LOADPARM. If no site ID is identified in the LOADPARM, then the
common namespace gets activated.
If some variable is missed (undefined) in the active namespace, then it
gets evaluated by the common namespace. If some variable is missed in
both, active and common namespaces, then it gets removed from the command
line by the boot process.

.SH OPTIONS
.TP
.BR "\-h" " or " "\-\-help"
Print usage information, then exit.

.TP
.BR "\-v" " or " "\-\-version"
Print version information, then exit.

.TP
.BR "\-t <TARGET DIRECTORY>" " or " "\-\-target <TARGET DIRECTORY>"
Specify a directory, where the environment is installed. This directory should
contain boot data (bootmap file). If this option is not specified, then the
tool assumes that the environment is installed in  "/boot". A similar option
with the same name exists also for
.B zipl(8)
utility.

.TP
.BR "\-l" " or " "\-\-list"
Prints a list of zIPL environment variables with their values as found in the
installed environment block.

In a combination with the option -S (--site) it prints only values defined in
the respective site namespace.
In a combination with the option -E (--effective-site) it prints the way of
evaluation of zIPL environment variables that would take place, if the
specified site was encoded in the LOADPARM for the boot session.
By default it simply dumps all the installed namespaces.

.TP
.BR "\-s <NAME=VALUE>" " or " "\-\-set <NAME=VALUE>"
Assign
.B VALUE
to the variable
.B NAME.
.B NAME
has to satisfy the requirements above.
.B VALUE
may consist of any printable characters different from the new line symbol.
If variable with such name didn't exist in the environment,
then it will be added.
In a combination with the option -S (--site) the value is assigned in the
specified namespace. By default - in the common namespace.
.TP
.BR "\-u <NAME>" " or " "\-\-unset <NAME>"
Remove the variable with name
.B NAME
from zIPL environment.
In a combination with the option -S (--site) the variable gets removed from
the specified namespace. By default - from the common namespace.
.TP
.BR "\-r" " or " "\-\-reset"
Remove all variables from zIPL environment.
In a combination with the option -S (--site) the variables get removed only
from the specified site. By default - from all namespaces.
.TP
.BR "\-S" " or " "\-\-site <SITE_ID>"
Specifies a particular site namespace to operate on. Can be used in a
combination with options \-s (\-\-set), \-u (\-\-unset), -r (\-\-reset) and
\-l (\-\-list).
This option makes the tool operate on a specified namespace only.
Specifically, when using in a combination with \-s, \-u, or -r, changes are
applied only to the specified namespace. When using in a combination with
\-l, only specified namespace is displayed.
.B SITE_ID
is a site identifier - any decimal positive number from 0 till 9.
.TP
.BR "\-E" " or " "\-\-effective\-site <SITE_ID>"
When using in a combination with option \-l (\-\-list), it displays the way
of evaluation of zIPL environment variables that would take place, if the
specified site was activated for the boot session.
.B SITE_ID
is a site identifier - any decimal positive number from 0 till 9.

.SH FILES

.B /etc/ziplenv
- "initial environment" (or "environment file").

A regular file containing zIPL environment to be installed every time when
preparing a device for initial program load (IPL) by
.B zipl(8)
utility.
Don't confuse environment file with environment block!
In contrast with environment block, environment file is not a boot component
and is not used at boot time by any boot processes.
This file contains pairs
.B NAME=VALUE
delimited by new line symbols (i.e. each pair on a separate line).
.B NAME
is identified as a sequence of characters before the leftmost '=' in the line.
.B NAME
has to satisfy the requirements above. Lines beginning from "#" are ignored.
Lines with not identified
.B NAME
are ignored.

Namespaces in an environment file are specified by sections. Sections are
indicated by titles - lines, starting with a keyword '[site X]', where X is a
namespace ID. Rectangle brackets in any such keyword are mandatory. The word
"site" is case insensitive. Any keyword may be followed with a comment on the
same line. Any lines between neighboring section titles do form a section,
defined by the upper title. If the environment file doesn't begin from a
section title, then the area at the beginning, not indicated by a title,
defines the "common" namespace. Sections with identical titles are considered
as parts of the same "compound" section. If a section (simple or compound)
contains lines with identical names, then only the last one takes an effect.

If the file defines more than 512 effective values (summed over all sections)
or if all namespaces defined by that file don't fit to the environment block,
then
.B zipl(8)
will fail to import such file.

This file can be modified, using any suitable text editor.

.SH NOTES

Site namespace corresponds to a
.B fail-over site
which is a set of block devices, participating in hardware replication,
and used for planned/unplanned swap.

Any installed zIPL environment (environment block) should be accessed only
by this
.B zipl-editenv
tool. Using other ways to access the environment block is not allowed.

Any modification of installed zIPL environment block doesn't affect the
environment file. The user is responsible for keeping the environment file
up to date.

.TP

.SH SEE ALSO
.BR zipl (8),
.BR zipl.conf (5)
.BR chreipl (8)
.BR lsreipl (8)


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------
* __v2.20.x (2022-xx-xx)__

  For Linux kernel version: 5.xx

  Add new tools / libraries:

  Changes of existing tools:
  - dbginfo.sh: rework script formatting (indents, order)
  - dbginfo.sh: update sysfs collection (excludes, messages)
  - dbginfo.sh: replace "which" by builtin command "type"

  Bug Fixes:

* __v2.20.0 (2022-02-04)__

  For Linux kernel version: 5.16

  Add new tools / libraries:
  - Add EditorConfig configuration

  Changes of existing tools:
  - s390-tools switches to Fuse 3 as Fuse 2 is deprecated.
        Affected tools: cmsfs, hmcdrvfs, hsavmcore, zdsfs, zdump
  - chreipl-fcp-mpath: don't compress the manpage before installing it
  - cpictl: Report extended version information
  - genprotimg: Add extended kernel command line support
  - zdev: modify the lsblk output parser in lszdev
  - zipl: Add support for longer kernel command lines (now supports up to 64k length)

  Bug Fixes:
  - cpictl: Suppress messages for unwritable sysfs files
  - dbginfo.sh: Fix missing syslog for step create_package
  - lshwc: Fix CPU list parameter setup for device driver
  - zdev: Check for errors when removing a devtype setting
  - zdev: Fix path resolution for multi-mount point file systems

* __v2.19.0 (2021-11-10)__

  For Linux kernel version: 5.15

  Add new tools / libraries:
  - chreipl-fcp-mpath: New toolset that uses multipath information to change
      the configured FCP re-IPL path on detecting issues with the current path

  Changes of existing tools:
  - dbginfo.sh: Add retry timeout and remove possible blocking "blockdev --report"
  - dbginfo.sh: Collect config- and debug-data for chreipl-fcp-mpath
  - hsci: Add support for multiple MAC addresses

  Bug Fixes:
  - lshwc: Fix compile error for gcc <8.1
  - zdump: Various clean-ups and fixes
  - ziomon: Correct throughput calculation in ziorep_printers
  - zipl: Fix segmentation fault when setting stage3_parms

* __v2.18.0 (2021-10-01)__

  For Linux kernel version: 5.14

  Add new tools:
  - scripts: Add tool for parsing sclp s390dbf logs
  - zdev: Add udev rule helper tool
  - zipl-editenv: Add tool to operate with zIPL environment installed in the boot record

  Changes of existing tools:
  - Makefile: Fix order of build of libraries for parallel builds
  - dbginfo.sh: Add collection in area of timedate, coredump and --check option
  - dbginfo.sh: Add exception on dump2tar for /sys/kernel/mm/page_idle/bitmap
  - dbginfo.sh: Cleanup of outdated sections and general code rework
  - dbginfo.sh: Collect zipl boot menu entries from boot loader specification
  - lszcrypt: Add support for vfio-ap status field
  - lszcrypt: Improved output for deconfig cards and queues
  - lszfcp: Add linkdown case to host marker of extended output
  - zdev: Add auto-config for PCI and crypto devices
  - zdump: Introduce multi-level message logging
  - zipl: Add support for environment block interpretation
  - zkey-cryptsetup: Support LUKS2 volumes with integrity support enabled

  Bug Fixes:
  - hsavmcore: Avoid recompilation of overlay during install step
  - libkmipclient: Fix parsing of hex values for XML and JSON encoding
  - vmur/vmur.cpp: Fix error handling on transfer failure
  - zdump: Lots of smaller fixes across the board

* __v2.17.0 (2021-07-07)__

  For Linux kernel version: 5.12 / 5.13

  Add new tools / libraries:
  - hsavmcore: New utility to make the dump process with kdump more efficient
  - libkmipclient: Add KMIP client shared library
  - libseckey: Add a secure key library
  - lshwc: New tool to extract and list complete counter sets

  Changes of existing tools:
  - genprotimg: Add '--(enable|disable)-pckmo' options
  - genprotimg: Add OpenSSL 3.0 support
  - genprotimg: Change plaintext control flags defaults so PCKMO functions are allowed
  - libutil: Introduce multi-level message logging (util_log)
  - libutil: Introduce util_arch module
  - udev/dasd: Change DASD udev-rule to set none scheduler
  - zdsfs: Add transparent codepage conversion
  - zkey: Add support for KMIP-based key management systems

  Bug Fixes:
  - ttyrun-getty: Avoid conflicts with serial-getty@
  - dbginfo: add /proc/kallsyms - refresh zVM, lscpu - fix WORKARCHIVE handling
  - dbginfo: add KVM data collection for server and guest - fix lszdev
  - genprotimg: Add missing return values in error paths
  - zkey: Fix conversion of CCA DATA keys to CCA CIPHER keys
  - znetconf: avoid conflict with "chzdev -e"

* __v2.16.0 (2021-02-19)__

  For Linux kernel version: 5.10 / 5.11

  Add new tool:
  - hsci: New tool to manage HSCI (HiperSockets Converged Interfaces)

  Changes of existing tools:
  - genprotimg: Add host-key document verification support
  - genprotimg: boot: Make boot loader -march=z900 compatible
  - libekmfweb: Make install directory for shared libraries configurable
  - lsdasd: Add FC Endpoint Security information
  - make: Add address sanitizer support
  - netboot: Add version information to scripts
  - netboot: Bump busybox version in pxelinux.0 build
  - zdev: Add FC Endpoint Security information for DASD devices
  - zdev: Add build option to update initial RAM-disk by default
  - zkey-ekmfweb: Avoid sequence number clash when generating keys
  - zkey/zkey-ekmfweb: Install KMS plugins into configurable location
  - zkey: Add support to store LUKS2 dummy passphrase in key repository

  Bug Fixes:
  - dasdfmt: Fix segfault when an incorrect option is specified
  - genprotimg: Fix several build issues
  - genprotimg: Require argument for 'ramdisk' and 'parmfile' options
  - zcryptstats: Fix handling of partial results with many domains
  - zfcpdbf: Deal with crash 7.2.9 change in caller name formatting
  - zipl/boot: Fix memory use after free in stage2
  - zipl/boot: Fix potential heap overflow in stage2
  - zipl: Fix reading 4k disk's geometry

* __v2.15.1 (2020-10-28)__

  For Linux kernel version: 5.9

  Changes of existing tools:

zipl/src/zipl-editenv.c
/*
 * zipl-editenv: Environment Editor for zSeries Initial Program Loader
 *
 * Copyright IBM Corp. 2020
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <assert.h>
#include "disk.h"
#include "error.h"
#include "misc.h"
#include "bootmap.h"
#include "envblk.h"
#include "lib/util_prg.h"
#include "lib/util_opt.h"
#include "lib/util_part.h"
#include "lib/util_path.h"
#include "lib/util_proc.h"

/* from linux/fs.h */
#define FIBMAP		_IO(0x00, 1)
#define FIGETBSZ	_IO(0x00, 2)

static const struct util_prg prg = {
	.desc = "zipl-editenv: Environment Editor for zSeries Initial Program Loader",
	.copyright_vec = {
		{
			.owner = "IBM Corp.",
			.pub_first = 2021,
		},
		UTIL_PRG_COPYRIGHT_END
	}
};

int verbose;
static char *bootmap_dir;

enum op_id {
	INVALID_OP_ID,
	SET_OP_ID,
	UNSET_OP_ID,
	RESET_OP_ID,
	LIST_OP_ID,
	LAST_OP_ID
};

static char *op_desc[LAST_OP_ID] = {"invalid", "set", "unset", "reset", "list"};

struct zipl_envblk {
	int fd;
	char *buf;
	off_t offset;
	int size;
	unsigned int need_update:1;
};

static char *opcode2desc(enum op_id opcode)
{
	return op_desc[opcode];
}

static void zipl_envblk_init(struct zipl_envblk *zeb)
{
	memset(zeb, 0, sizeof(*zeb));
	zeb->fd = -1;
}

static void print_name_value(char *this)
{
	printf("%s\n", this);
}

static int do_list(char *envblk, unsigned int envblk_size)
{
	return envblk_scan(envblk, envblk_size, print_name_value);
}

static int do_set(char *envblk, unsigned int envblk_size,
		  const char *name, const char *new_val)
{
	unsigned int name_len, new_val_len;
	char *fss; /* free space start */
	unsigned int lines_scanned = 0;
	int name_found = 0;
	char *s, *end;

	name_len = strlen(name);
	new_val_len = strlen(new_val);

	s = envblk + sizeof(ZIPL_ENVBLK_SIGNATURE) - 1;
	end = envblk + envblk_size;

	/*
	 * find the start of free space
	 */
	for (fss = end - 1; *fss == '\0'; fss--)
		;
	if (*fss != '\n') {
		error_reason("Found corrupted environment block - please run zipl");
		return -1;
	}
	fss++;

	while (fss - s > name_len) {
		if (lines_scanned >= ENVBLK_MAX_LINES) {
			error_reason("Found corrupted environment block - please run zipl");
			return -1;
		}
		if (memcmp(s, name, name_len) == 0 && s[name_len] == '=') {
			unsigned int cur_val_len;
			/*
			 * such name exists, replace its current value
			 */
			s += (name_len + 1);

			cur_val_len = 0;
			while (s + cur_val_len < end && s[cur_val_len] != '\n')
				cur_val_len++;
			if (s + cur_val_len >= end) {
				error_reason("Found corrupted environment block - please run zipl");
				return -1;
			}
			if (new_val_len > cur_val_len &&
			    end - fss < new_val_len - cur_val_len) {
				error_reason("Not enough space for new value");
				return -1;
			}
			/*
			 * make a precise-sized room for the new value
			 */
			if (new_val_len < cur_val_len) {
				memmove(s + new_val_len, s + cur_val_len,
					end - (s + cur_val_len));

				memset(fss + cur_val_len - new_val_len, '\0',
				       cur_val_len - new_val_len);
			} else
				memmove(s + new_val_len, s + cur_val_len,
					end - (s + new_val_len));
			name_found = 1;
			break;
		}
		lines_scanned++;
		s = envblk_next_line(s, end);
	}
	assert(lines_scanned <= ENVBLK_MAX_LINES);

	if (!name_found) {
		if (lines_scanned == ENVBLK_MAX_LINES) {
			error_reason("Maximum number of lines reached");
			return -1;
		}
		/*
		 * append a new variable
		 */
		if (end - fss < name_len + new_val_len + 2) {
			error_reason("Not enough space in environment block");
			return -1;
		}
		memcpy(fss, name, name_len);
		s = fss + name_len;
		*s++ = '=';
	}
	/*
	 * copy the new value and terminate it with a new line symbol
	 */
	memcpy(s, new_val, new_val_len);
	s[new_val_len] = '\n';
	return 0;
}

static int do_unset(char *envblk, int envblk_len, const char *name)
{
	unsigned int name_len;
	char *s, *end;

	name_len = strlen(name);
	s = envblk + sizeof(ZIPL_ENVBLK_SIGNATURE) - 1;
	end = envblk + envblk_len;

	while (end - s >= name_len + 2 /* minimal length of
					* pattern "name=foo"
					*/) {
		if (memcmp(s, name, name_len) == 0 && s[name_len] == '=') {
			/*
			 * name was found. Locate the whole
			 * "named" line and cut it including
			 * the trailing "\n"
			 */
			unsigned int cut_len = name_len + 1;

			while (s + cut_len < end) {
				if (s[cut_len] == '\n')
					break;
				cut_len++;
			}
			if (s + cut_len >= end) {
				/*
				 * trailing "\n" not found
				 */
				error_reason("Found corrupted environment block - please run zipl");
				return -1;
			}
			cut_len++;
			memmove(s, s + cut_len, end - (s + cut_len));
			memset(end - cut_len, '\0', cut_len);
			return 0;
		}
		s = envblk_next_line(s, end);
	}
	error_reason("Name %s not found", name);
	return -1;
}

/**
 * sync the content of the environment block with disk
 */
static int envblk_update(struct zipl_envblk *zeb)
{
	struct util_proc_part_entry part_entry;
	struct stat info;
	blocknum_t blknr;
	char *dev_name;
	int dev_fd;

	if (fstat(zeb->fd, &info))
		return -1;

	if (util_proc_part_get_entry(info.st_dev, &part_entry) != 0)
		return -1;

	dev_name = misc_make_path("/dev", part_entry.name);
	if (!dev_name) {
		error_reason("Could not make path for %s", part_entry.name);
		return -1;
	}
	dev_fd = open(dev_name, O_RDWR | O_DIRECT);
	if (dev_fd < 0) {
		error_reason("Could not open file %s", dev_name);
		free(dev_name);
		goto error;
	}
	free(dev_name);
	/*
	 * det disk address of the environment block
	 */
	if (fs_map(zeb->fd, zeb->offset, &blknr, zeb->size) != 0)
		goto error_close;

	if (lseek64(dev_fd, blknr * (uint64_t)zeb->size, SEEK_SET) < 0) {
		error_reason(strerror(errno));
		goto error_close;
	}
	if (write(dev_fd, zeb->buf, zeb->size) != zeb->size) {
		error_reason("Could not update physical block %ld on %s",
			     blknr, part_entry.name);
		goto error_close;
	}
	if (verbose)
		printf("Physical block %ld on %s got updated\n",
		       blknr, part_entry.name);

	util_proc_part_free_entry(&part_entry);
	return 0;
error_close:
	close(dev_fd);
error:
	util_proc_part_free_entry(&part_entry);
	return -1;
}

static int envblk_close(struct zipl_envblk *zeb)
{
	if (zeb->buf != NULL) {
		free(zeb->buf);
		zeb->buf = NULL;
	}
	if (zeb->fd < 0)
		return 0;

	if (close(zeb->fd) == 0) {
		zeb->fd = -1;
		return 0;
	}
	return -1;
}

/**
 * Open a regular file, which contains environment block, and
 * read that block to a buffer. Fill in environment block info
 * ZEB with valid values
 */
static int envblk_open(struct zipl_envblk *zeb)
{
	char *bootmap_file;

	if (zeb->fd >= 0)
		/*  it was opened before */
		return 0;
	bootmap_file =
		misc_make_path(bootmap_dir != NULL ? bootmap_dir : "/boot",
			       "bootmap");
	if (!bootmap_file) {
		error_reason("Could not make path for bootmap file");
		return -1;
	}
	zeb->fd = open(bootmap_file, O_RDONLY);
	if (zeb->fd < 0) {
		error_reason("Could not open bootmap file %s: %s",
			     bootmap_file, strerror(errno));
		free(bootmap_file);
		bootmap_file = NULL;
		return -1;
	}
	if (verbose)
		printf("Processing bootmap file at %s\n", bootmap_file);

	if (envblk_size_get(zeb->fd, &zeb->size)) {
		error_reason("Could not get environment block size");
		goto error;
	}
	if (envblk_offset_get(zeb->fd, &zeb->offset)) {
		error_reason("Could not get environment block location");
		goto error;
	}
	if (zeb->offset == 0) {
		error_reason("Environment not installed - please run zipl");
		goto error;
	}
	/* reopen for direct operations */
	close(zeb->fd);
	zeb->fd = open(bootmap_file, O_RDWR | O_DIRECT);
	if (zeb->fd < 0) {
		error_reason("Could not open environment block at %s",
			     bootmap_file);
		goto error;
	}
	free(bootmap_file);
	bootmap_file = NULL;

	zeb->buf = aligned_alloc(zeb->size, zeb->size);
	if (!zeb->buf) {
		error_reason("Could not allocate aligned memory region");
		goto error;
	}
	if (lseek(zeb->fd, zeb->offset, SEEK_SET) < 0) {
		error_reason(strerror(errno));
		goto error;
	}
	if (read(zeb->fd, zeb->buf, zeb->size) != zeb->size) {
		error_reason("Could not read environment block");
		goto error;
	}
	if (memcmp(zeb->buf, ZIPL_ENVBLK_SIGNATURE,
		sizeof(ZIPL_ENVBLK_SIGNATURE) - 1) != 0) {
		error_reason("Found corrupted environment block - please run zipl");
		goto error;
	}
	return 0;
error:
	if (bootmap_file)
		free(bootmap_file);
	envblk_close(zeb);
	return -1;
}

static int env_list(struct zipl_envblk *zeb)
{
	if (envblk_open(zeb))
		return -1;
	return do_list(zeb->buf, strnlen(zeb->buf, zeb->size));
}

static int env_set(struct zipl_envblk *zeb, char *arg)
{
	char *name, *value;

	if (envblk_open(zeb))
		return -1;
	/*
	 * parse the argument, locate name and value
	 */
	name = strdup(arg);
	if (!name)
		return -1;

	value = strchr(name, '=');
	if (!value || value == name) {
		/* name is not identified */
		error_reason("Invalid argument %s", arg);
		goto error;
	}
	*value = '\0';
	value++;

	if (envblk_check_name(name, strlen(name))) {
		error_reason("Unacceptable name '%s'", name);
		goto error;
	}
	if (do_set(zeb->buf, zeb->size, name, value) == 0) {
		free(name);
		zeb->need_update = 1;
		return 0;
	}
error:
	free(name);
	return -1;
}

static int env_unset(struct zipl_envblk *zeb, char *name)
{
	if (envblk_open(zeb))
		return -1;
	/*
	 * check name for validness
	 */
	if (strchr(name, '=') != NULL) {
		error_reason("Invalid argument %s", name);
		return -1;
	}
	if (do_unset(zeb->buf, strnlen(zeb->buf, zeb->size), name) == 0) {
		zeb->need_update = 1;
		return 0;
	}
	return -1;
}

static void env_blank(char *envblk, int envblk_len)
{
	memset(envblk + sizeof(ZIPL_ENVBLK_SIGNATURE) - 1, '\0',
	       envblk_len - sizeof(ZIPL_ENVBLK_SIGNATURE) + 1);
}

static int env_reset(struct zipl_envblk *zeb)
{
	if (envblk_open(zeb))
		return -1;

	env_blank(zeb->buf, zeb->size);
	zeb->need_update = 1;
	return 0;
}

static int save_op(enum op_id *dst, enum op_id src)
{
	if (*dst != INVALID_OP_ID) {
		error_reason("Can't perform more than one operation (%s and %s) at a time\n",
			     opcode2desc(*dst), opcode2desc(src));
		return -EINVAL;
	}
	*dst = src;
	return 0;
}

static int save_op_arg(enum op_id *dst, enum op_id src,
		       char **dst_arg, char *src_arg)
{
	int ret;

	ret = save_op(dst, src);
	if (ret)
		return ret;
	*dst_arg = strdup(src_arg);
	if (*dst_arg == NULL) {
		error_reason("Couldn't copy the argument\n");
		return -ENOMEM;
	}
	return 0;
}

static struct util_opt opt_vec[] = {
	UTIL_OPT_SECTION("OPTIONS WITHOUT ARGUMENTS"),
	{
		.option = { "list", no_argument, NULL, 'l'},
		.desc = "print list of zIPL environment variables with their values",
	},
	{
		.option = { "reset", no_argument, NULL, 'r'},
		.desc = "remove all variables from zIPL environment",
	},
	{
		.option = { "verbose", no_argument, NULL, 'V'},
		.desc = "provide more information",
	},
	UTIL_OPT_SECTION("OPTIONS WITH ARGUMENTS"),
	{
		.option = { "target", required_argument, NULL, 't'},
		.argument = "DIR",
		.desc = "specify directory, where bootmap file is located",
	},
	{
		.option = { "set", required_argument, NULL, 's'},
		.argument = "NAME=VALUE",
		.desc = "assign value VALUE to variable NAME",
	},
	{
		.option = { "unset", required_argument, NULL, 'u'},
		.argument = "NAME",
		.desc = "remove variable NAME from zIPL environment",
	},
	UTIL_OPT_SECTION("STANDARD OPTIONS"),
	UTIL_OPT_HELP,
	UTIL_OPT_VERSION,
	UTIL_OPT_END,
};

int main(int argc, char *argv[])
{
	enum op_id opcode = INVALID_OP_ID;
	struct zipl_envblk zeb;
	char *saved_arg = NULL;
	int ret = 0;
	int c;

	zipl_envblk_init(&zeb);
	util_prg_init(&prg);
	util_opt_init(opt_vec, NULL);

	if (argc < 2) {
		fprintf(stderr,
			"%s: Missing command\nUse 'zipl-editenv --help' for more information\n",
			argv[0]);
		return EXIT_FAILURE;
	}
	while (1) {
		c = util_opt_getopt_long(argc, argv);
		if (c == -1)
			break;
		switch (c) {
		case 'h':
			util_prg_print_help();
			util_opt_print_help();
			goto out;
		case 'v':
			util_prg_print_version();
			goto out;
		case 'V':
			verbose = 1;
			break;
		case 't':
			bootmap_dir = optarg;
			break;
		case 'r':
			ret = save_op(&opcode, RESET_OP_ID);
			break;
		case 'l':
			ret = save_op(&opcode, LIST_OP_ID);
			break;
		case 's':
			ret = save_op_arg(&opcode, SET_OP_ID,
					  &saved_arg, optarg);
			break;
		case 'u':
			ret = save_op_arg(&opcode, UNSET_OP_ID,
					  &saved_arg, optarg);
			break;
		default:
			util_opt_print_parse_error(c, argv);
			ret = -1;
		}
		if (ret)
			goto out;
	}
	if (optind < argc) {
		fprintf(stderr, "%s: Unknown command", argv[0]);
		while (optind < argc)
			printf(" %s", argv[optind++]);
		fprintf(stderr, "\n");
		goto out;
	}
	switch (opcode) {
	case SET_OP_ID:
		ret = env_set(&zeb, saved_arg);
		break;
	case UNSET_OP_ID:
		ret = env_unset(&zeb, saved_arg);
		break;
	case RESET_OP_ID:
		ret = env_reset(&zeb);
		break;
	case LIST_OP_ID:
		ret = env_list(&zeb);
		break;
	default:
		break;
	}
out:
	if (zeb.need_update)
		ret = envblk_update(&zeb);

	envblk_close(&zeb);
	if (saved_arg != NULL)
		free(saved_arg);
	if (ret) {
		error_print();
		return EXIT_FAILURE;
	}
	return EXIT_SUCCESS;
}
